new-tutorial/L1_large_node_classification.py

```diff
@@ -0,0 +1,326 @@
+valid_nids = idx_split['valid']
+test_nids = idx_split['test']
+
+
+######################################################################
+# How DGL Handles Computation Dependency
+# --------------------------------------
+# 
+# In the :doc:`previous tutorial <L0_neighbor_sampling_overview>`, you
+# have seen that the computation dependency for message passing of a
+# single node can be described as a series of bipartite graphs.
+# 
+# |image1|
+# 
+# .. |image1| image:: https://data.dgl.ai/tutorial/img/bipartite.gif
+# 
+
+
+######################################################################
+# Defining Neighbor Sampler and Data Loader in DGL
+# ------------------------------------------------
+# 
+# DGL provides useful tools to iterate over the dataset in minibatches
+# while generating the computation dependencies to compute their outputs
+# with the bipartite graphs above. For node classification, you can use
+# ``dgl.dataloading.NodeDataLoader`` for iterating over the dataset. Then
+# you can use ``dgl.dataloading.MultiLayerNeighborSampler`` to generate
+# computation dependencies of the nodes from a multi-layer GNN with
+# *neighbor sampling*, i.e. taking only a fixed number of neighbors for
+# each node to aggregate messages.
+# 
+# The syntax of ``dgl.dataloading.NodeDataLoader`` is mostly similar to a
+# PyTorch ``DataLoader``, with the addition that it needs a graph to
+# generate computation dependency from, a set of node IDs to iterate on,
+# and the neighbor sampler you defined.
+# 
+# Let’s say that each node will gather message from 4 neighbors on each
+# layer. The code defining the data loader and neighbor sampler will look
+# like the following.
+# 
+
+sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+train_dataloader = dgl.dataloading.NodeDataLoader(
+    graph, train_nids, sampler,
+    batch_size=1024,
+    shuffle=True,
+    drop_last=False,
+    num_workers=0,
+    device='cuda'
+)
+
+
+######################################################################
+# You can iterate over the data loader and see what it yields.
+# 
+
+input_nodes, output_nodes, bipartites = example_minibatch = next(iter(train_dataloader))
+print(example_minibatch)
+print("To compute {} nodes' outputs, we need {} nodes' input features".format(len(output_nodes), len(input_nodes))) 
+
+
+######################################################################
+# ``NodeDataLoader`` gives us three items per iteration.
+# 
+# -  The input node list for the nodes whose input features are needed to
+#    compute the outputs.
+# -  The output node list whose GNN representation are to be computed.
+# -  The list of computation dependency for each layer as a list of
+#    bipartite graphs.
+# 
+
+
+######################################################################
+# You can get the input and output node IDs of the first bipartite graph
+# and see that the first few input nodes are always the same as the output
+# nodes.
```

Comment:
". As we described in the previous overview page, output nodes' own previous features may also be necessary in the computation of their new features"